src/integrations/pulsar/consumer.cpp

```diff
@@ -0,0 +1,273 @@
+// Copyright 2021 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "integrations/pulsar/consumer.hpp"
+
+#include <fmt/format.h>
+#include <pulsar/Client.h>
+
+#include <chrono>
+#include <thread>
+
+#include "integrations/constants.hpp"
+#include "integrations/pulsar/exceptions.hpp"
+#include "utils/concepts.hpp"
+#include "utils/logging.hpp"
+#include "utils/on_scope_exit.hpp"
+#include "utils/result.hpp"
+#include "utils/thread.hpp"
+
+namespace integrations::pulsar {
+
+namespace {
+utils::BasicResult<std::string, std::vector<Message>> GetBatch(pulsar_client::Consumer &consumer,
+                                                               const ConsumerInfo &info, std::atomic<bool> &is_running,
+                                                               std::optional<uint64_t> &next_message_timestamp) {
+  std::vector<Message> batch{};
+
+  const auto batch_size = info.batch_size.value_or(kDefaultBatchSize);
+  batch.reserve(batch_size);
+
+  auto remaining_timeout_in_ms = info.batch_interval.value_or(kDefaultBatchInterval).count();
+  auto start = std::chrono::steady_clock::now();
+
+  bool run_batch = true;
+  for (int64_t i = 0; remaining_timeout_in_ms > 0 && i < batch_size && is_running.load(); ++i) {
+    pulsar_client::Message message;
+    const auto result = consumer.receive(message, static_cast<int>(remaining_timeout_in_ms));
+    switch (result) {
+      case pulsar_client::Result::ResultTimeout:
+        run_batch = false;
+        break;
+      case pulsar_client::Result::ResultOk:
+        if (next_message_timestamp) {
+          if (*next_message_timestamp == message.getPublishTimestamp()) {
+            batch.emplace_back(Message{std::move(message)});
+            next_message_timestamp.reset();
+          }
+        } else {
+          batch.emplace_back(Message{std::move(message)});
+        }
```

Comment:
I need a comment for this, but basically, we are ignoring all the messages until we get the one that should be next.
While I ask Pulsar to resend the unacked messages, new ones can get in in the same time, basically duplicating them.
I.e. you get 6, 7, 0, 1, 2, 3, 4, 5, 6, 7...